storage/kubernetes/controller.go

```diff
@@ -160,7 +163,28 @@ func (s *storage) targetSecret() (*v1.Secret, error) {
 }
 
 func (s *storage) saveInK8s(secret *v1.Secret) (*v1.Secret, error) {
+
 	if !s.initComplete() {
+		// Start a goroutine to attempt to save the secret later, once init is complete.
+		// If this was already handled by initComplete, it should be a no-op, or at worst get
+		// merged with the Kubernetes secret.
+		go s.initSync.Do(func() {
+			err := wait.Poll(100*time.Millisecond, 15*time.Minute, func() (bool, error) {
+				if !s.initComplete() {
+					return false, nil
+				}
+				secret, err := s.saveInK8s(secret)
+				if err != nil {
+					logrus.Errorf("failed to save TLS secret for %s/%s: %v", secret.Namespace, secret.Name, err)
+					return false, err
+				}
+				return true, nil
+			})
+
+			if err != nil {
+				logrus.Errorf("error while waiting for storage to initialized to save secret %s/%s: %v", secret.Namespace, secret.Name, err)
```

Comment:
1. The standard form for this when not returning is the more compact `if err := foo(); err != nil {`
2. Don't store the secret returned from the inner `saveInK8s` call; we're not going to do anything with it and it shadows the `secret` value from the outer scope.
3. Don't log the error from the inner `saveInK8s` twice, just return it and let it be logged as the err return from `Poll`
4. Log messages should be capitalized like a sentence.
```suggestion
			if err := wait.Poll(100*time.Millisecond, 15*time.Minute, func() (bool, error) {
				if !s.initComplete() {
					return false, nil
				}
				_, err := s.saveInK8s(secret)
				return true, err
			}); err != nil {
				logrus.Warnf("Failed to save TLS secret after controller init: %v", err)
```

Reply:
updated.